packages/astroport-governance/src/assembly.rs

```diff
@@ -201,6 +245,59 @@ pub struct Proposal {
     pub deposit_amount: Uint128,
 }
 
+impl Proposal {
+    pub fn validate(&self, whitelisted_links: Vec<String>) -> StdResult<()> {
+        // Title validation
+        if self.title.len() < MIN_TITLE_LENGTH {
+            return Err(StdError::generic_err("Title too short!"));
+        }
+
+        if self.title.len() > MAX_TITLE_LENGTH {
+            return Err(StdError::generic_err("Title too long!"));
+        }
+        if !self.description.chars().all(char::is_alphanumeric) {
+            return Err(StdError::generic_err(
+                "Title is not in alphanumeric format!",
+            ));
+        }
+
+        // Description validation
+        if self.description.len() < MIN_DESC_LENGTH {
+            return Err(StdError::generic_err("Description too short!"));
+        }
+        if self.description.len() > MAX_DESC_LENGTH {
+            return Err(StdError::generic_err("Description too long!"));
+        }
+        if !self.description.chars().all(char::is_alphanumeric) {
+            return Err(StdError::generic_err(
+                "Description is not in alphanumeric format",
+            ));
+        }
+
+        // Link validation
+        if let Some(link) = &self.link {
+            if link.len() < MIN_LINK_LENGTH {
+                return Err(StdError::generic_err("Link too short!"));
+            }
+            if link.len() > MAX_LINK_LENGTH {
+                return Err(StdError::generic_err("Link too long!"));
+            }
+            if !whitelisted_links.iter().any(|wl| link.starts_with(wl)) {
+                return Err(StdError::generic_err("Link is not whitelisted!"));
+            }
+            // Avoid using unsafe characters
+            if !link
+                .chars()
+                .all(|c| char::is_ascii(&c) && !['<', '>', ' ', '\\', '{', '}'].contains(&c))
```

Comment:
This constant array ```['<', '>', ' ', '\\', '{', '}']``` will be created every time when the closure launches. I suggest to create it before 289 line.

Reply:
It makes sense. Changed, thanks.

Reply:
It makes sense. Changed, thanks.